File: apps/desktop/src/i18n.ts
import localeRegistryJson from './locales/registry.generated.json'

export type AppLanguage = string
export type TranslationTree = string | TranslationTree[] | { [key: string]: TranslationTree }
export type BreakIdeaEntry = {
  title: string | null
  text: string
}

export type AppLanguageConfig = {
  code: string
  label: string
  nativeLabel: string
  fallback: string
  direction: 'ltr' | 'rtl'
  desktopReady: boolean
}

type LocaleRegistry = {
  defaultLanguage: string
  languages: AppLanguageConfig[]
  bundles: Record<string, TranslationTree>
}

const localeRegistry = localeRegistryJson as LocaleRegistry
const languageConfigs = new Map(
  localeRegistry.languages.map((config) => [config.code, config] as const),
)
const defaultLanguage = localeRegistry.defaultLanguage
const defaultConfig = languageConfigs.get(defaultLanguage) ?? localeRegistry.languages[0]

export const AVAILABLE_LANGUAGE_CONFIGS = [...localeRegistry.languages]
export const DESKTOP_LANGUAGE_CONFIGS = [...AVAILABLE_LANGUAGE_CONFIGS].sort((left, right) => {
  if (left.code === defaultLanguage) {
    return -1
  }
  if (right.code === defaultLanguage) {
    return 1
  }
  return left.label.localeCompare(right.label, 'en')
})

function canonicalLanguage(language?: string): string | undefined {
  switch (language) {
    case 'zh':
    case 'zh_CN':
    case 'zh-Hans':
    case 'zh-Hans-CN':
      return 'zh-CN'
    default:
      return language
  }
}

function rawConfig(language?: string): AppLanguageConfig {
  const canonical = canonicalLanguage(language)
  return (canonical ? languageConfigs.get(canonical) : undefined) ?? defaultConfig
}

function resolvedConfig(language?: string): AppLanguageConfig {
  return rawConfig(language)
}

function bundleChain(language?: string): TranslationTree[] {
  const chain: TranslationTree[] = []
  const seen = new Set<string>()
  let current: AppLanguageConfig | undefined = rawConfig(language)

  while (current && !seen.has(current.code)) {
    seen.add(current.code)
    const bundle = localeRegistry.bundles[current.code]
    if (bundle) {
      chain.push(bundle)
    }
    current = current.fallback ? languageConfigs.get(current.fallback) : undefined
  }

  const resolved = resolvedConfig(language)
  if (!seen.has(resolved.code)) {
    const bundle = localeRegistry.bundles[resolved.code]
    if (bundle) {
      chain.push(bundle)
    }
  }

  return chain.length > 0 ? chain : [localeRegistry.bundles[defaultLanguage]]
}

export function getLanguageConfig(language?: string): AppLanguageConfig {
  return resolvedConfig(language)
}

export function normalizeLanguage(language?: string): AppLanguage {
  return resolvedConfig(language).code
}

export function resolveUiLanguage(
  draftLanguage?: AppLanguage,
  persistedLanguage?: AppLanguage,
): AppLanguage {
  return normalizeLanguage(persistedLanguage ?? draftLanguage)
}

function lookup(bundle: TranslationTree, key: string): TranslationTree | null {
  let current: TranslationTree | undefined = bundle

  for (const segment of key.split('.')) {
    if (!current || typeof current === 'string' || Array.isArray(current)) {
      return null
    }
    current = current[segment]
  }

  return current ?? null
}

export function t(
  language: AppLanguage,
  key: string,
  vars: Record<string, string | number> = {},
): string {
  for (const bundle of bundleChain(language)) {
    const template = lookup(bundle, key)
    if (typeof template !== 'string') {
      continue
    }

    return Object.entries(vars).reduce((result, [name, value]) => {
      return result.split(`{{${name}}}`).join(String(value))
    }, template)
  }

  return key
}

export function tList(language: AppLanguage, key: string): string[] {
  for (const bundle of bundleChain(language)) {
    const value = lookup(bundle, key)
    if (Array.isArray(value) && value.every((entry) => typeof entry === 'string')) {
      return [...value]
    }
  }

  return []
}

export function tBreakIdeaList(
  language: AppLanguage,
  kind: 'microbreak' | 'longBreak',
): string[] {
  return tBreakIdeaEntries(language, kind).map((entry) => entry.text)
}

export function tBreakIdeaEntries(
  language: AppLanguage,
  kind: 'microbreak' | 'longBreak',
): BreakIdeaEntry[] {
  const key = kind === 'microbreak' ? 'miniBreakIdeas' : 'longBreakIdeas'

  for (const bundle of bundleChain(language)) {
    const value = lookup(bundle, key)
    if (!value || typeof value === 'string' || Array.isArray(value)) {
      continue
    }

    const prompts = Object.values(value)
      .map((entry) => {
        if (!entry || typeof entry === 'string' || Array.isArray(entry)) {
          return null
        }

        const text = entry.text
        if (typeof text !== 'string') {
          return null
        }

        return {
          title: typeof entry.title === 'string' ? entry.title : null,
          text,
        }
      })
      .filter((entry): entry is BreakIdeaEntry => Boolean(entry))

    if (prompts.length > 0) {
      return prompts
    }
  }

  return []
}

export function formatDuration(ms: number | null, language: AppLanguage): string {
  const normalizedLanguage = normalizeLanguage(language)

  if (!ms || ms <= 0) {
    return normalizedLanguage === 'zh-CN' ? '0分' : '0m'
  }

  const totalSeconds = Math.ceil(ms / 1000)
  const minutes = Math.floor(totalSeconds / 60)
  const seconds = totalSeconds % 60

  if (minutes === 0) {
    return t(normalizedLanguage, 'duration.seconds', { value: seconds })
  }

  if (seconds === 0) {
    return t(normalizedLanguage, 'duration.minutes', { value: minutes })
  }

  const paddedSeconds =
    normalizedLanguage === 'zh-CN'
      ? seconds.toString().padStart(2, '0')
      : `${seconds}s`

  return t(normalizedLanguage, 'duration.minutesSeconds', {

    minutes,
    seconds: paddedSeconds,
  })
}

export function formatCountdown(ms: number | null): string {
  const totalSeconds = Math.max(0, Math.ceil((ms ?? 0) / 1000))
  const hours = Math.floor(totalSeconds / 3600)
  const minutes = Math.floor((totalSeconds % 3600) / 60)
  const seconds = totalSeconds % 60

  if (hours > 0) {
    return [hours, minutes, seconds].map((value) => value.toString().padStart(2, '0')).join(':')
  }

  return [minutes, seconds].map((value) => value.toString().padStart(2, '0')).join(':')
}

File: apps/desktop/src-tauri/src/i18n.rs
use serde::Deserialize;
use serde_json::Value;
use std::collections::{HashMap, HashSet};
use std::sync::OnceLock;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct LanguageConfig {
    code: String,
    fallback: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct LocaleRegistry {
    default_language: String,
    languages: Vec<LanguageConfig>,
    bundles: HashMap<String, Value>,
}

static REGISTRY: OnceLock<LocaleRegistry> = OnceLock::new();

pub fn normalize_language(language: &str) -> &'static str {
    let Some(default) = default_config() else {
        return "zh-CN";
    };

    config_for_code(canonical_language(language))
        .unwrap_or(default)
        .code
        .as_str()
}

pub fn text(language: &str, key: &str) -> String {
    template(language, key)
}

pub fn text1(language: &str, key: &str, name: &str, value: impl AsRef<str>) -> String {
    template(language, key).replace(&placeholder(name), value.as_ref())
}

pub fn text2(
    language: &str,
    key: &str,
    first_name: &str,
    first_value: impl AsRef<str>,
    second_name: &str,
    second_value: impl AsRef<str>,
) -> String {
    text1(language, key, first_name, first_value.as_ref())
        .replace(&placeholder(second_name), second_value.as_ref())
}

pub fn duration(language: &str, ms: u64) -> String {
    let normalized_language = normalize_language(language);
    let total_seconds = (ms / 1000).max(1);
    let minutes = total_seconds / 60;
    let seconds = total_seconds % 60;

    if minutes == 0 {
        return text1(normalized_language, "duration.seconds", "value", seconds.to_string());
    }

    if seconds == 0 {
        return text1(normalized_language, "duration.minutes", "value", minutes.to_string());
    }

    let seconds_value = if normalized_language == "zh-CN" {
        format!("{seconds:02}")
    } else {
        format!("{seconds}s")
    };

    text2(
        normalized_language,
        "duration.minutesSeconds",
        "minutes",
        minutes.to_string(),
        "seconds",
        seconds_value,
    )
}

fn registry() -> &'static LocaleRegistry {
    REGISTRY.get_or_init(|| {
        serde_json::from_str(include_str!("../../src/locales/registry.generated.json"))
            .expect("valid locale registry")
    })
}

fn config_for_code(code: &str) -> Option<&'static LanguageConfig> {
    registry().languages.iter().find(|config| config.code == code)
}

fn default_config() -> Option<&'static LanguageConfig> {
    config_for_code(&registry().default_language)
        .or_else(|| registry().languages.first())
}

fn canonical_language(language: &str) -> &str {
    match language {
        "zh" | "zh_CN" | "zh-Hans" | "zh-Hans-CN" => "zh-CN",
        _ => language,
    }
}

fn bundle_chain(language: &str) -> Vec<&'static Value> {
    let registry = registry();
    let Some(default) = default_config() else {
        return Vec::new();
    };

    let mut chain = Vec::new();
    let mut seen = HashSet::new();
    let mut current = Some(config_for_code(canonical_language(language)).unwrap_or(default));

    while let Some(config) = current {
        if !seen.insert(config.code.as_str()) {
            break;
        }

        if let Some(bundle) = registry.bundles.get(&config.code) {
            chain.push(bundle);
        }

        current = if config.fallback.is_empty() {
            None
        } else {
            config_for_code(&config.fallback)
        };
    }

    let normalized = normalize_language(language);
    if seen.insert(normalized) {
        if let Some(bundle) = registry.bundles.get(normalized) {
            chain.push(bundle);
        }
    }

    chain
}

fn placeholder(name: &str) -> String {
    format!("{{{{{name}}}}}")
}

fn template(language: &str, key: &str) -> String {
    for bundle in bundle_chain(language) {
        if let Some(template) = lookup(bundle, key) {
            return template;
        }
    }

    key.to_string()
}

fn lookup(bundle: &Value, key: &str) -> Option<String> {
    let mut current = bundle;

    for segment in key.split('.') {
        current = current.get(segment)?;
    }

    current.as_str().map(ToOwned::to_owned)
}
